### PR TITLE
Spec: Clarify identity partition edge cases. 

### PR DESCRIPTION
**Pull request by emkornfield**
Discussion on mailing list: https://lists.apache.org/thread/hss83r1605r8932b94xv9y2wfb9o0yns.

### CODE REVIEW

**Comment by emkornfield** [comment]
CC @rdblue @RussellSpitzer an alternative (or perhaps cleanup) would be to move column projection to be a subject of scan-planning.  I think this would flow more nicely since all of the concepts would have been introduced already (instead of having to forward link)

File: format/spec.md [hunk 1]
```diff
@@ -399,6 +401,9 @@ Sorting floating-point numbers should produce the following behavior: `-NaN` < `
 
 A data or delete file is associated with a sort order by the sort order's id within [a manifest](#manifests). Therefore, the table must declare all the sort orders for lookup. A table could also be configured with a default sort order id, indicating how the new data should be sorted by default. Writers should use this default sort order to sort the data on write, but are not required to if the default order is prohibitively expensive, as it would be for streaming writes.
 
+#### Writing with Identity transform
+
+When writing data files, all columns including those with an identity transforms should be written to data files. This provides redundancy in case of corruption or bugs in the metadata layer. Due to [column projection rules](#column-projection) readers can still properly scan the table if columns that have an indentity partition transforms applied are ommitted. This is not the  case for any other transform type.
```
Comment (ajantha-bhat) [inline format/spec.md:406]:
```suggestion
When writing data files, all columns including those with an identity transforms should be written to data files. This provides redundancy in case of corruption or bugs in the metadata layer. Due to [column projection rules](#column-projection) readers can still properly scan the table if columns that have an identity partition transforms applied are omitted. This is not the case for any other transform type.
```

File: format/spec.md [hunk 2]
```diff
@@ -241,7 +241,9 @@ Struct evolution requires the following rules for default values:
 
 #### Column Projection
 
-Columns in Iceberg data files are selected by field id. The table schema's column names and order may change after a data file is written, and projection must be done using field ids. If a field id is missing from a data file, its value for each row should be `null`.
+Columns in Iceberg data files are selected by field id. The table schema's column names and order may change after a data file is written, and projection must be done using field ids.
+
+When a projected column has an [identity partition transform](#partition-transforms) applied to it for a data file, the value from the [manifest file](#manifests) must be used for that column (i.e. the column should not be read from the data file). This is to support tables that were migrated from other table formats (notably Hive) that do not write partition values to data files. Otherwise, if a field id is missing from a data file, its value for each row should be `null`.
```
Comment (ajantha-bhat) [inline format/spec.md:246]: Question, if projected column has other partition transform, then also the value must be read from manifest file itself right? I am not clear on why this is specific to identity partition transform?

Comment (RussellSpitzer) [inline format/spec.md:246]: Isn't this no longer going to be true with default values as well? If a field id is missing the value should be "null" unless a default value is specified.

Comment (RussellSpitzer) [inline format/spec.md:246]: I'm not sure I understand the wording here.

Can we be more explicit? Something like :

"Values for Field Ids which are not present in a data file **must** be projected as 
  1. The value in the partition metadata if an Identity Transform exists for that field or
  2. The default value as defined in [Default values} or
  3. null
  
  
  ^ Just a thought though, I'm open to other wordings or presentations

Comment (ajantha-bhat) [reply format/spec.md:246]: So, we can update "identity partition transform" ->  "partition transform" and remove mention about hive and consider default values. 

How about like this?

> When a projected column has [partition transform] (#partition-transforms) applied to it for a data file, the value from the [manifest file](#manifests) must be used for that column (i.e. the column should not be read from the data file). Otherwise, if a field id is missing from a data file, its value for each row should be `null` unless a default value is specified.

File: format/spec.md [hunk 3]
```diff
@@ -399,6 +401,9 @@ Sorting floating-point numbers should produce the following behavior: `-NaN` < `
 
 A data or delete file is associated with a sort order by the sort order's id within [a manifest](#manifests). Therefore, the table must declare all the sort orders for lookup. A table could also be configured with a default sort order id, indicating how the new data should be sorted by default. Writers should use this default sort order to sort the data on write, but are not required to if the default order is prohibitively expensive, as it would be for streaming writes.
 
+#### Writing with Identity transform
+
+When writing data files, all columns including those with an identity transforms should be written to data files. This provides redundancy in case of corruption or bugs in the metadata layer. Due to [column projection rules](#column-projection) readers can still properly scan the table if columns that have an identity partition transforms applied are omitted. This is not the case for any other transform type.
```
Comment (RussellSpitzer) [inline format/spec.md:406]: I'm not sure the "Due to" sentence is helpful here. I'm generally in favor of trying to keep the spec as sparse as possible.


"When writing data files,"  can probably be removed from the first sentence as well without changing the meaning of the sentence.

I'm not sure "those with an identity transform" is the right description. "All columns, including those whose values are also present in partition metadata, should be written to datafiles"?

Comment (emkornfield) [reply format/spec.md]: Rephrased and moved this to implementation notes.

File: format/spec.md (hunk 2 above)
Comment (emkornfield) [reply format/spec.md]: Used a modified version, of this let me know if it makes sense to you.

Comment (emkornfield) [reply format/spec.md]: changed wording as suggested by @RussellSpitzer

File: format/spec.md [hunk 4]
```diff
@@ -241,7 +241,13 @@ Struct evolution requires the following rules for default values:
 
 #### Column Projection
 
-Columns in Iceberg data files are selected by field id. The table schema's column names and order may change after a data file is written, and projection must be done using field ids. If a field id is missing from a data file, its value for each row should be `null`.
+Columns in Iceberg data files are selected by field id. The table schema's column names and order may change after a data file is written, and projection must be done using field ids.
+
+Values for Field ids which are not present in a data file must be resolved according the following rules:
```
Comment (rdblue) [inline format/spec.md:246]: Typo: `Field` should be `field`

File: format/spec.md [hunk 5]
```diff
@@ -241,7 +241,13 @@ Struct evolution requires the following rules for default values:
...
+
+* Return the value from partition metadata if an [Identity Transform](#partition-transforms) exists for the field.
```
Comment (rdblue) [inline format/spec.md:248]: > Return the value from partition metadata if the data file's partition metadata contains the field as a result of the `identity` transform.

I want to be clear that this is only required if the data file's partition metadata contains the field, not just if there is a partition spec that uses an identity transform for the field.

File: format/spec.md [hunk 6]
```diff
@@ -591,11 +597,10 @@ For example, an `events` table with a timestamp column named `ts` that is partit
 
 Scan predicates are also used to filter data and delete files using column bounds and counts that are stored by field id in manifests. The same filter logic can be used for both data and delete files because both store metrics of the rows either inserted or deleted. If metrics show that a delete file has no rows that match a scan predicate, it may be ignored just as a data file would be ignored [2].
 
-Data files that match the query filter must be read by the scan. 
+Data files that match the query filter must be read by the scan.
 
 Note that for any snapshot, all file paths marked with "ADDED" or "EXISTING" may appear at most once across all manifest files in the snapshot. If a file path appears more than once, the results of the scan are undefined. Reader implementations may raise an error in this case, but are not required to do so.
 
-
```
Comment (rdblue) [inline format/spec.md:598]: Nit: unnecessary whitespace change.

File: format/spec.md [hunk 7]
```diff
@@ -1393,4 +1398,8 @@ This section covers topics not required by the specification but recommendations
...
\ No newline at end of file
+When processing point in time queries implementations should use "snapshot-log" metadata to lookup the table state at the given point in time. This ensures time-travel queries reflect the state of the table at the provided timestamp. For example a SQL query like `SELECT * FROM prod.db.table TIMESTAMP AS OF '1986-10-26 01:21:00Z';` would find the snapshot of the Iceberg table just prior to '1986-10-26 01:21:00 UTC' in the snapshot logs and use the metadata from that snapshot to perform the scan of the table. If no  snapshot exists prior to the timestamp given or "snapshot-log" is not populated (it is an optional field), then systems should raise an informative error message about the missing metadata.
+
+### Writing data files
+
+All columns should be written to data files even if they introduce redundancy with metadata stored in manifest file (e.g. columns with identity partition transforms). Writing all columns provides redundancy in case of corruption or bugs in the metadata layer.
```
Comment (RussellSpitzer) [inline format/spec.md:1405]: nit: Second redundancy could be "a back up"

**Review by RussellSpitzer — APPROVED** [review]

File: format/spec.md (hunk 7 above)
Comment (rdblue) [inline format/spec.md:1405]: I think that this should be in the main spec. Writers are required to write all columns, but readers must handle cases where columns are missing. Those are real requirements, not conventions.

File: format/spec.md [hunk 8]
```diff
@@ -591,11 +597,10 @@ For example, an `events` table with a timestamp column named `ts` that is partit
 
 Scan predicates are also used to filter data and delete files using column bounds and counts that are stored by field id in manifests. The same filter logic can be used for both data and delete files because both store metrics of the rows either inserted or deleted. If metrics show that a delete file has no rows that match a scan predicate, it may be ignored just as a data file would be ignored [2].
 
-Data files that match the query filter must be read by the scan. 
+Data files that match the query filter must be read by the scan.
```
Comment (rdblue) [inline format/spec.md:600]: Nit: unnecessary whitespace change.

File: format/spec.md [hunk 9]
```diff
@@ -241,7 +241,13 @@ Struct evolution requires the following rules for default values:
 
 #### Column Projection
 
-Columns in Iceberg data files are selected by field id. The table schema's column names and order may change after a data file is written, and projection must be done using field ids. If a field id is missing from a data file, its value for each row should be `null`.
+Columns in Iceberg data files are selected by field id. The table schema's column names and order may change after a data file is written, and projection must be done using field ids.
+
+Values for Field ids which are not present in a data file must be resolved according the following rules:
+
+* Return the value from partition metadata if an [Identity Transform](#partition-transforms) exists for the field. 
+* Return the default value as defined in [Default values](#default-values) if it exists.
```
Comment (rdblue) [inline format/spec.md:249]: I think we can be clear here that in cases where the column is not present in a data file (after applying name mappings) the `initial-default` must be used.

File: format/spec.md (hunk 7 above)
Comment (RussellSpitzer) [reply format/spec.md:1405]: Let's change "should" to "must" instead then.

Comment (emkornfield) [reply format/spec.md]: changed.  Moved this above

Comment (emkornfield) [reply format/spec.md]: changed.

File: format/spec.md (hunk 9 above)
Comment (emkornfield) [reply format/spec.md]: I split the name mapping into a different step.

File: format/spec.md (hunk 5 above)
Comment (emkornfield) [reply format/spec.md]: I think I addressed this but not sure if I fully understood the comment.

File: format/spec.md (hunk 4 above)
Comment (emkornfield) [reply format/spec.md]: done.

File: format/spec.md [hunk 10]
```diff
@@ -241,7 +245,14 @@ Struct evolution requires the following rules for default values:
 
 #### Column Projection
 
-Columns in Iceberg data files are selected by field id. The table schema's column names and order may change after a data file is written, and projection must be done using field ids. If a field id is missing from a data file, its value for each row should be `null`.
+Columns in Iceberg data files are selected by field id. The table schema's column names and order may change after a data file is written, and projection must be done using field ids.
+
+Values for field ids which are not present in a data file must be resolved according the following rules:
+
+* Return the value from partition metadata if an [Identity Transform](#partition-transforms) exists for the field and the partition value is present in the `partitition` struct on  `data_file` object in the manifest.
```
Comment (ajantha-bhat) [inline format/spec.md:252]:
```suggestion
* Return the value from partition metadata if an [Identity Transform](#partition-transforms) exists for the field and the partition value is present in the `partition` struct on `data_file` object in the manifest. 
```

**Review by ajantha-bhat — APPROVED** [review]

**Review by jackye1995 — APPROVED** [review]

File: format/spec.md [hunk 11]
```diff
@@ -241,7 +245,14 @@ Struct evolution requires the following rules for default values:
...
 
-Columns in Iceberg data files are selected by field id. The table schema's column names and order may change after a data file is written, and projection must be done using field ids. If a field id is missing from a data file, its value for each row should be `null`.
+Columns in Iceberg data files are selected by field id. The table schema's column names and order may change after a data file is written, and projection must be done using field ids.
+
+Values for field ids which are not present in a data file must be resolved according the following rules:
+
+* Return the value from partition metadata if an [Identity Transform](#partition-transforms) exists for the field and the partition value is present in the `partition` struct on `data_file` object in the manifest. 
+* Use `schema.name-mapping.default` metadata to map field id to columns without field id as described below and use the column if it is present.
+* Return the default value if it has a defined in `initial-default` (See [Default values](#default-values) section for more details). 
+* Return `null` in all other cases.
```
Comment (singhpk234) [inline format/spec.md:255]: [minor] should we move these lines after we explain what `schema.name-mapping.default` is ?

**Review by singhpk234 — COMMENTED** [review]
Thanks @emkornfield for taking this up. 

This is super helpful in streamlining what we need to do hive tables migrated to iceberg tables would need to do as the data files didn't persist partition col in the file itself as hive catalog tracks it. 

One had to read the code to understand how iceberg-core does this handling with all fallback and can be easily missed when doing diff lang impl, I came across it when i was adding iceberg support for redshift and had to go through the similar exercise.

File: format/spec.md (hunk 11 above)
Comment (emkornfield) [reply format/spec.md:255]: I referenced that it is described below.  I placed it i moved text from the following paragraph and highlights the overall logic is non-trivial.

**Review by Fokko — APPROVED** [review]
Thanks for clearing this up 👍

**Review by rdblue — APPROVED** [review]

**Review by amogh-jahagirdar — APPROVED** [review]

File: format/spec.md [hunk 12]
```diff
@@ -241,7 +245,14 @@ Struct evolution requires the following rules for default values:
 
 #### Column Projection
 
-Columns in Iceberg data files are selected by field id. The table schema's column names and order may change after a data file is written, and projection must be done using field ids. If a field id is missing from a data file, its value for each row should be `null`.
+Columns in Iceberg data files are selected by field id. The table schema's column names and order may change after a data file is written, and projection must be done using field ids.
+
+Values for field ids which are not present in a data file must be resolved according the following rules:
```
Comment (findepi) [inline format/spec.md:250]: Above says "All columns must be written to data file".
Here it says "field ids which are not present in a data file ..."
Besides schema evolution (new columns) case,  and multiple partitionings in use, when can the field ids  be missing from the data file?

File: format/spec.md [hunk 13]
```diff
@@ -241,7 +245,14 @@ Struct evolution requires the following rules for default values:
...
+
+* Return the value from partition metadata if an [Identity Transform](#partition-transforms) exists for the field and the partition value is present in the `partition` struct on `data_file` object in the manifest.
```
Comment (findepi) [inline format/spec.md:252]: if a new column projection is defined (eg `day(order_time)`), can the query engine derive value for that field from the `order_time` column?

File: format/spec.md (hunk 12 above)
Comment (rdblue) [reply format/spec.md:250]: When files are added from Hive tables or other places. If you're writing to an Iceberg table we have stricter standards than if you are adding data files from a converted table.

In the past, we didn't document the expectations for readers, but this is definitely a case where we want to be clear about the requirements across engines.

Comment (emkornfield) [reply format/spec.md:250]: When migrating from other table formats that don't write this information into data files.  Hive is the primary example.

**Review by emkornfield — COMMENTED** [review]
add back reasoning for partition columns.

File: format/spec.md (hunk 13 above)
Comment (rdblue) [reply format/spec.md:252]: Transforms are one way, with the exception of `identity` that does not modify the value. No other partition transforms support recovering the original value.

If you have a case where the engine is deriving a value that is stored in metadata (which I _think_ is what you're asking) then you can use the metadata value as long as you know that the transform the engine is projecting exactly matches the Iceberg transform.

File: format/spec.md (hunk 12 above)
Comment (findepi) [reply format/spec.md:250]: I get it -- files migrated from hive lack some information and we need to deal with it.
If such "partial data files" are legal, why do we expect writers not to create them? The readers still need to support them, so what's the win (from format perspective)?

File: format/spec.md (hunk 13 above)
Comment (emkornfield) [reply format/spec.md:252]: I'm not sure I fully understand the question but if order_time is of type date, then in this special case the transform is effectively an identity transform.

File: format/spec.md (hunk 12 above)
Comment (emkornfield) [reply format/spec.md:250]: The argument here is that especially with default values there is a larger room for misinterpretation.  I don't have a strong feeling here.

File: format/spec.md (hunk 13 above)
Comment (findepi) [reply format/spec.md]: I was thinking about situation like this (using Trino syntax)

```
-- create a table with some data
CREATE TABLE t AS SELECT 123 AS a;

-- add new partitioning column
ALTER TABLE t SET PROPERTIES partitioning = ARRAY['truncate(a, 10)'];
```

Now we have two fields: the `a` data column and `a_trunc` projected column.
Trino doesn't provide a way to query for `a_trunc` column directly.
However, if it did, the value for `a_trunc` could be derived from the data.

Comment (emkornfield) [reply format/spec.md]: It seems for this case we probably want a subsection of the specification or an implementation note for projecting partition values?  Maybe that would be a good follow-up (I'm not sure which engines actually allow projecting partition values since they aren't technically part of the schema.)

Comment (findepi) [reply format/spec.md]: Yes, agreed that partition columns are not part of the schema, so don't show up in table's relational model.
However, the document here talks about reading "field ids" and partition columns are also fields with ids. Or is this section supposed to be only about fields that are plain data fields?

Comment (emkornfield) [reply format/spec.md]: I read it as at least initially only about plain data, since for transform partition values these would never actually be in data file, and that is what this topic currently discusses.  I think adding a subsection about projecting partition values if necessary to this section as a follow-up could make sense.

File: format/spec.md (hunk 12 above)
Comment (rdblue) [reply format/spec.md:250]: Files are the source of truth for row data. It's convenient to have a copy of identity columns in metadata because we can avoid projecting and doing extra work. But you should still get a full copy of the row as it was written if you read just the file.

File: format/spec.md (hunk 13 above)
Comment (rdblue) [reply format/spec.md]: This is allowed, particularly for filters where you can use Iceberg's expression library to filter by partition values directly. You can also use partition values in aggregations (which is under development).

I don't think that changes what is said here. In this case, the value from metadata should be used if it is present.

Comment (findepi) [reply format/spec.md]: > the value from metadata should be used if it is present.

makes sense, agreed

but if it's not present -- why require to return null? we could require the query engine to compute the value based on the source columns present in the data file

Comment (rdblue) [reply format/spec.md]: @findepi, the spec originally stated that all columns that were not present in a file must be interpreted as `null` because the file is the source of truth for row data -- if a column was not present then it must not have existed at the time the file was written and must default to `null` because that was the only possible default for new columns.

This is how identity partition columns were always required to be written because omitting them would create a conflict between the partition value in metadata and the source of truth in the file. Now, the requirement to write all columns is more explicitly stated above.

Over time, we added support for reading Hive files that were missing the identity partitioned columns. This now documents how to do that in the read path but we didn't drop the requirement to write all columns in the write path.

In addition, we've now added column defaults for v3 and didn't realize that this section conflicted with the `initial-default` behavior.

This is just stating more clearly what the expected behavior is and I think it's a good change because it is clear.

**Review by danielcweeks — APPROVED** [review]

**Comment by emkornfield** [comment]
[Vote passed](https://lists.apache.org/thread/44s6vkd9ky6x4ty2fyn9j7r7ylry6lgh)

**Comment by emkornfield** [comment]
Could a committer merge?
